datalad/distribution/publish.py

```diff
@@ -256,49 +282,90 @@ def __call__(path=None, dataset=None, to=None, recursive=False, since=None,
                 raise ValueError("No sibling '%s' found." % dest_resolved)
 
-            lgr.info("Publishing {0} to {1}".format(ds, dest_resolved))
-
-            # we now know where to push to:
-            _log_push_info(ds.repo.push(remote=dest_resolved,
-                                        refspec=ds.repo.get_active_branch(),
-                                        set_upstream=set_upstream))
+            # in order to be able to use git's config to determine what to push,
+            # we need to annex merge first. Otherwise a git push might be
+            # rejected if involving all matching branches for example.
+            # Once at it, also push the annex branch right here.
 
-            # push annex branch:
+            # Q: Do we need to respect annex-ignore here? Does it make sense to
+            # publish to a remote without pushing the annex branch
+            # (if there is any)?
             if isinstance(ds.repo, AnnexRepo):
                 ds.repo.fetch(remote=dest_resolved)
                 ds.repo.merge_annex(dest_resolved)
                 _log_push_info(ds.repo.push(remote=dest_resolved,
                                             refspec="git-annex:git-annex"))
 
-            # we need to fetch
-            # TODO
-            # Note: This is about a gitpython issue as well as something about
-            # annex -> might mean, that we need to do it in case we pushed an annex branch only.
-            # Apparently, we can annex copy new files only, after this fetch. Figure it out!
-            ds.repo.fetch(remote=dest_resolved)
+            # upstream branch needed for update (merge) and subsequent push,
+            # in case there is no.
+            # no tracking branch yet?
+            set_upstream = track_branch is None
+
+            # publishing of `dest_resolved` might depend on publishing other
+            # remote(s) first:
+            # define config var name for potential publication dependencies
+            depvar = 'remote.{}.datalad-publish-depends'.format(dest_resolved)
+            for d in ds.config.get(depvar, []):
+                lgr.info("Dependency detected: '%s'" % d)
+                # Note: Additional info on publishing the dep. comes from within
+                # `ds.publish`.
+                ds.publish(path=path,
+                           to=d,
+                           since=since,
+                           skip_failing=skip_failing,
+                           recursive=recursive,
+                           recursion_limit=recursion_limit,
+                           git_opts=git_opts,
+                           annex_opts=annex_opts,
+                           annex_copy_opts=annex_copy_opts)
+
+            lgr.info("Publishing {0} to {1}".format(ds, dest_resolved))
+
+            # we now know where to push to:
+            # TODO: what to push? default: git push --mirror if nothing configured?
+            # consider also: --follow-tags, --tags, --atomic
```

Comment:
as we discussed during one of the hangouts before (see in notes from 09/15/2016), and I thought agreed upon, best would be just to rely on settings within `.git/config`.  That would eliminate such questions here in publish.  There is really no need to reinvent the wheel/duplicate settings within publish or what could be set directly within git.  That would then "natively" provide resolution to #481 (if git config was setup accordingly by e.g. create-sibling, see #925)